win32/devdoc/socket_transport_win32.md

```diff
@@ -0,0 +1,296 @@
+    SOCKET_SEND_ERROR, \
+    SOCKET_SEND_SHUTDOWN
+
+MU_DEFINE_ENUM(SOCKET_SEND_RESULT, SOCKET_SEND_RESULT_VALUES)
+
+#define SOCKET_RECEIVE_RESULT_VALUES \
+    SOCKET_RECEIVE_OK, \
+    SOCKET_RECEIVE_WOULD_BLOCK, \
+    SOCKET_RECEIVE_ERROR, \
+    SOCKET_RECEIVE_SHUTDOWN
+
+MU_DEFINE_ENUM(SOCKET_RECEIVE_RESULT, SOCKET_RECEIVE_RESULT_VALUES)
+
+#define SOCKET_TYPE_VALUES \
+    SOCKET_CLIENT, \
+    SOCKET_SERVER
+
+MU_DEFINE_ENUM(SOCKET_TYPE, SOCKET_TYPE_VALUES)
+
+typedef struct SOCKET_BUFFER_TAG
+{
+    uint32_t length;
+    void* buffer;
+} SOCKET_BUFFER;
+
+MOCKABLE_FUNCTION(, SOCKET_TRANSPORT_HANDLE, socket_transport_create, SOCKET_TYPE, type);
+MOCKABLE_FUNCTION(, void, socket_transport_destroy, SOCKET_TRANSPORT_HANDLE, socket_transport);
+
+MOCKABLE_FUNCTION(, int, socket_transport_connect, SOCKET_TRANSPORT_HANDLE, socket_transport, const char*, hostname, uint16_t, port, uint32_t, connection_timeout_ms);
+MOCKABLE_FUNCTION(, int, socket_transport_listen, SOCKET_TRANSPORT_HANDLE, socket_transport, uint16_t, port);
+MOCKABLE_FUNCTION(, void, socket_transport_disconnect, SOCKET_TRANSPORT_HANDLE, socket_transport);
+
+MOCKABLE_FUNCTION(, SOCKET_TRANSPORT_HANDLE, socket_transport_accept, SOCKET_TRANSPORT_HANDLE, socket_transport);
+
+MOCKABLE_FUNCTION(, SOCKET_SEND_RESULT, socket_transport_send, SOCKET_TRANSPORT_HANDLE, socket_transport, SOCKET_BUFFER*, payload, uint32_t, buffer_count, uint32_t*, bytes_written, uint32_t, flags, void*, data);
+MOCKABLE_FUNCTION(, SOCKET_RECEIVE_RESULT, socket_transport_receive, SOCKET_TRANSPORT_HANDLE, socket_transport, SOCKET_BUFFER*, payload, uint32_t, buffer_count, uint32_t*, bytes_recv, uint32_t, flags, void*, data);
+
+MOCKABLE_FUNCTION(, SOCKET_HANDLE, socket_transport_get_underlying_socket, SOCKET_TRANSPORT_HANDLE, socket_transport);
+```
+
+### socket_transport_create
+
+```c
+MOCKABLE_FUNCTION(, SOCKET_TRANSPORT_HANDLE, socket_transport_create, SOCKET_TYPE, type);
+```
+
+`socket_transport_create` creates a socket transport.
+
+`socket_transport_create` shall allocate a new `SOCKET_TRANSPORT` object.
+
+`socket_transport_create` shall call `sm_create` to create a sm object.
+
+On any failure `socket_transport_create` shall return `NULL`.
+
+On success `socket_transport_create` shall return `SOCKET_TRANSPORT_HANDLE`.
+
+### socket_transport_destroy
+
+```c
+MOCKABLE_FUNCTION(, void, socket_transport_destroy, SOCKET_TRANSPORT_HANDLE, socket_transport);
+```
+
+`socket_transport_destroy` destroys all data stored in the `SOCKET_TRANSPORT_HANDLE` object.
+
+If `socket_transport` is `NULL` `socket_transport_destroy` shall do nothing.
+
+`socket_transport_destroy` shall call `sm_destroy` to destroy the sm object.
+
+`socket_transport_destroy` shall free the `SOCKET_TRANSPORT_HANDLE` object.
+
+### socket_transport_connect
+
+```c
+MOCKABLE_FUNCTION(, int, socket_transport_connect, SOCKET_TRANSPORT_HANDLE, socket_transport, const char*, hostname, uint16_t, port, uint32_t, connection_timeout_ms);
```

Comment:
>connection_timeout_ms [](http://example.com/codeflow?start=143&length=21)

not clear what happnes with this parameter

Reply:
I will add a spec for it.